python/iceberg/parquet/parquet_reader.py

```diff
@@ -0,0 +1,240 @@
+                                              for nested_field in field.type.fields]), {}),
+     TypeID.TIMESTAMP: lambda field: (pa.timestamp("us"), datetime.now()),
+     # not used in SPARK, so not implementing for now
+     # TypeID.TIME: pa.time64(None)
+     }
+
+
+class ParquetReader(object):
+
+    def __init__(self, input: InputFile, expected_schema: Schema, options, filter_expr: Expression,
+                 case_sensitive: bool, start: int = None, end: int = None):
+        self._stats: Dict[str, int] = dict()
+
+        self._input = input
+        self._input_fo = input.new_fo()
+
+        self._arrow_file = pq.ParquetFile(self._input_fo)
+        self._file_schema = convert_parquet_to_iceberg(self._arrow_file)
+        self._expected_schema = expected_schema
+        self._file_to_expected_name_map = ParquetReader.get_field_map(self._file_schema,
+                                                                      self._expected_schema)
+        self._options = options
+        self._filter = get_dataset_filter(filter_expr, ParquetReader.get_reverse_field_map(self._file_schema,
+                                                                                           self._expected_schema))
+
+        self._case_sensitive = case_sensitive
+        if start is not None or end is not None:
+            raise NotImplementedError("Partial file reads are not yet supported")
+            # self.start = start
+            # self.end = end
+
+        self.materialized_table = False
+        self.curr_iterator = None
+        self._table = None
+        self._df = None
+
+        _logger.debug("Reader initialized for %s" % self._input.path)
+
+    @property
+    def stats(self) -> dict:
+        return dict(self._stats)
+
+    def to_pandas(self) -> Union[pd.Series, pd.DataFrame]:
+        if not self.materialized_table:
+            self._read_data()
+            with profile("to_pandas", self._stats):
+                if self._table is not None:
+                    self._df = self._table.to_pandas(use_threads=True)
+                else:
+                    self._df = None
+
+        return self._df
+
+    def to_arrow_table(self) -> pa.Table:
+        if not self.materialized_table:
+            self._read_data()
+
+        return self._table
+
+    def _read_data(self) -> None:
+        _logger.debug("Starting data read")
+
+        # only scan the columns projected and in our file
+        cols_to_read = prune_columns(self._file_schema, self._expected_schema)
+
+        with profile("read data", self._stats):
+            arrow_dataset = ds.FileSystemDataset.from_paths([self._input.location()],
+                                                            schema=self._arrow_file.schema_arrow,
+                                                            format=ds.ParquetFileFormat(),
+                                                            filesystem=fs.LocalFileSystem())
```

Comment:
Is `LocalFileSystem` correct here? It caught my eye and wasn't sure how we would read eg from S3 or hadoop

Reply:
Let me change this to map from the filesystem type of the input files.  We always bring files local before reading because we have some very fast libs for sucking down s3 data. Much better than throughput than reading via boto, alhough I haven't benchmarked against what pyarrow is doing. We typically do not see a lot of row-group elimination. so the mileage on doing partial s3 reads is not that significant.  This may not be the general-case though, so it should probably pick a filesystem based on the input file.